crypto/stark-verifier-ethereum/contracts/fri.sol

```diff
@@ -29,20 +83,203 @@ contract Fri {
+
+    // This function takes in a previous layer and fold and reads from it and writes new folded layers to the next layer.
+    // It will overwrite any memory in that location.
+    function fold_layer(
+        uint256[] memory previous_layer,
+        uint64[] memory previous_indicies,
+        Iterators.IteratorUint memory extra_coset_data,
+        Eval_X memory eval_x,
+        uint256 eval_point,
+        LayerContext memory layer_context,
+        bytes32[] memory coset_hash_output
+    ) internal {
+        // Reads how many of the cosets we've read from
+        uint256 writes = 0;
+        uint64 current_index;
+        uint256[] memory next_coset = new uint256[](layer_context.coset_size);
+        uint256 i = 0;
+        while (i < previous_layer.length) {
+            current_index = previous_indicies[i];
+            // Each coset length elements in the domain are one coset, so to find which one the current index is
+            // we have to take it mod the length, to find the starting index we subtract the coset index from the
+            // current one.
+            uint64 min_coset_index = uint64((current_index) - (current_index % layer_context.coset_size));
+            for (uint64 j = 0; j < layer_context.coset_size; j++) {
+                // This check is if the current index is one which has data from the previous layer,
+                // or if it's one with data provided in the proof
+                if (current_index == j + min_coset_index) {
+                    // Set this coset's data to the previous layer data at this index
+                    next_coset[uint256(j)] = previous_layer[i];
+                    // Advance the index from the read
+                    i++;
+                    if (i < previous_indicies.length) {
+                        // Set the current index to the next one
+                        current_index = previous_indicies[i];
+                    }
+                } else {
+                    // This happens if the data isn't in the previous layer so we use our extra data.
+                    next_coset[uint256(j)] = extra_coset_data.next();
+                }
+            }
+            // Hash the coset and store it so we can do a merkle proof against it
+            coset_hash_output[writes] = merkleLeafHash(next_coset);
+            // Do the actual fold and write it to the next layer
+            // TODO - Mystery factor of two??
+            previous_layer[writes] = fold_coset(next_coset, eval_point, layer_context, min_coset_index / 2, eval_x);
+            // Record the new index
+            previous_indicies[writes] = uint64(min_coset_index / layer_context.coset_size);
+            writes++;
+        }
+        // We need to manually resize the output arrays;
+        assembly {
+            mstore(previous_layer, writes)
+            mstore(previous_indicies, writes)
+        }
+    }
+
+    function fold_coset(
+        uint256[] memory coset,
+        uint256 eval_point,
```

Comment:
We should probably come up with a better name than `eval_point`, but it matches the Rust implementation, so let's leave it for now.